src/data/rat/meta.lean

```diff
@@ -0,0 +1,99 @@
+/-
+Copyright (c) 2019 Robert Y. Lewis . All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Robert Y. Lewis
+-/
+
+import data.rat.basic
+
+/-!
+# Meta operations on ℚ
+
+This file defines functions for dealing with rational numbers as expressions.
+
+They are not defined earlier in the hierarchy, in the `tactic` or `meta` folders, since we do not
+want to import `data.rat.basic` there.
+
+## Main definitions
+
+* `rat.mk_numeral` embeds a rational `q` as a numeral expression into a type supporting the needed
+  operations. It does not need a tactic state.
+* `rat.reflect` specializes `rat.mk_numeral` to `ℚ`.
+* `expr.of_rat` behaves like `rat.mk_numeral`, but uses the tactic state to infer the needed
+  structure on the target type.
+
+* `expr.to_rat` evaluates a normal numeral expression as a rat.
+* `expr.eval_rat` evaluates a numeral expression with arithmetic operations as a rat.
+
+-/
+
+/--
+`rat.mk_numeral q` embeds `q` as a numeral expression inside a type with 0, 1, +, -, and /
+
+`type`: an expression representing the target type. This must live in Type 0.
+`has_zero`, `has_one`, `has_add`: expressions of the type `has_zero %%type`, etc.
+
+This function is similar to `expr.of_rat` but takes more hypotheses and is not tactic valued.
+ -/
+meta def rat.mk_numeral (type has_zero has_one has_add has_neg has_div : expr) : ℚ → expr
+| ⟨num, denom, _, _⟩ :=
+  let nume := num.mk_numeral type has_zero has_one has_add has_neg in
+  if denom = 1 then nume else
+    let dene := denom.mk_numeral type has_zero has_one has_add in
+    `(@has_div.div.{0} %%type %%has_div %%nume %%dene)
+
+/-- `rat.reflect q` represents the rational number `q` as a numeral expression of type `ℚ`. -/
+protected meta def rat.reflect : ℚ → expr :=
+rat.mk_numeral `(ℚ) `((by apply_instance : has_zero ℚ))
+         `((by apply_instance : has_one ℚ))`((by apply_instance : has_add ℚ))
+         `((by apply_instance : has_neg ℚ)) `(by apply_instance : has_div ℚ)
+
+section
+local attribute [semireducible] reflected
+meta instance : has_reflect ℚ := rat.reflect
+end
+
+/-- Evaluates an expression as a rational number,
+if that expression represents a numeral or the quotient of two numerals. -/
+protected meta def expr.to_pos_rat : expr → option ℚ
```

Comment:
Maybe this is too picky, but shouldn't this be `to_nonneg_rat`?

Reply:
This function isn't new, it was just moved from `norm_num`. But done.